extension/training/pybindings/_training_module.py

```diff
@@ -41,20 +41,20 @@ def forward_backward(self, method_name: str, inputs: Sequence[Any]) -> List[Any]
             self.parameters_method_prefix + method_name, ()
         )[0]
 
-        full_outputs = self.model.run_method(method_name, inputs)
+        # Important that the outputs are not cloned because we need the optimizer to
+        # be able to mutate the actual weights and not clones of them.
+        full_outputs = self.model.run_method(method_name, inputs, clone_outputs=False)
 
         user_outs = full_outputs[:grad_start_idx]
+        user_outs = [x.clone() for x in user_outs]
         grads = full_outputs[grad_start_idx:params_start_idx]
-        params = full_outputs[params_start_idx:]
+        grads = [grad.clone() for grad in grads]
 
-        # Important that the outputs are not cloned because we need the optimizer to
-        # be able to mutate the actual weights and not clones of them.
-        fqn = self.model.run_method(
-            self.fqn_method_prefix + method_name, (), clone_outputs=False
-        )
+        fqn = self.model.run_method(self.fqn_method_prefix + method_name, ())
 
         self.named_grads = dict(zip(fqn, grads))
         if self.named_params is None:
+            params = full_outputs[params_start_idx:]
```

Comment:
nit: would be more clear if you added the comment about not cloning here instead imo